### PR TITLE
Add better support for Raw Nodes in MFS and elsewhere

### PR DESCRIPTION
**Pull request by kevina**

### CODE REVIEW

**Comment by kevina** [comment]
Closes #3989

**Comment by kevina** [comment]
Note this is a work in progress.  In particular the Size() method in DagModifier needs to implemented to support raw nodes with MFS.  I am trying to figure out if it is possible to get the equivalent of Filesize from the generic Node interface.

**Comment by kevina** [comment]
Okay this is no longer a work in progress.  Fully generalize this to allow any type of node to be stored in the MFS tree should probably me another P.R.

**Comment by Kubuxu** [comment]
Could you add test cases from this PR: https://github.com/ipfs/go-ipfs/pull/3901/files

This makes sure that we test as much with RawNodes as we do test with normal ones.

File: unixfs/mod/dagmodifier.go [hunk 1]
```diff
@@ -301,13 +320,26 @@ func (dm *DagModifier) modifyDag(node *mdag.ProtoNode, offset uint64, data io.Re
...
 // appendData appends the blocks from the given chan to the end of this dag
-func (dm *DagModifier) appendData(node *mdag.ProtoNode, spl chunk.Splitter) (node.Node, error) {
+func (dm *DagModifier) appendData(nd node.Node, spl chunk.Splitter) (node.Node, error) {
+
+	var root *mdag.ProtoNode
+	switch nd := nd.(type) {
+	case *mdag.ProtoNode:
+		root = nd
+	case *mdag.RawNode:
+		// TODO: be able to append to rawnodes. Probably requires making this
```
Comment (Stebalien) [inline unixfs/mod/dagmodifier.go]: In this case, you're appending a raw node *to* `dm.curNode` which may or may not be raw.

File: unixfs/mod/dagmodifier.go [hunk 2]
```diff
@@ -301,13 +320,26 @@ func (dm *DagModifier) modifyDag(node *mdag.ProtoNode, offset uint64, data io.Re
 }
 
 // appendData appends the blocks from the given chan to the end of this dag
-func (dm *DagModifier) appendData(node *mdag.ProtoNode, spl chunk.Splitter) (node.Node, error) {
+func (dm *DagModifier) appendData(nd node.Node, spl chunk.Splitter) (node.Node, error) {
+
+	var root *mdag.ProtoNode
```
Comment (Stebalien) [inline unixfs/mod/dagmodifier.go]: Calling this `root` is misleading (it's the node to-be-appended).

File: unixfs/mod/dagmodifier.go [hunk 3]
```diff
@@ -301,13 +320,26 @@ func (dm *DagModifier) modifyDag(node *mdag.ProtoNode, offset uint64, data io.Re
...
+		return nil, ErrNotUnixfs
+	}
+
 	dbp := &help.DagBuilderParams{
 		Dagserv:  dm.dagserv,
 		Maxlinks: help.DefaultLinksPerBlock,
 	}
 
-	return trickle.TrickleAppend(dm.ctx, node, dbp.New(spl))
+	return trickle.TrickleAppend(dm.ctx, root, dbp.New(spl))
```
Comment (Stebalien) [inline unixfs/mod/dagmodifier.go]: The error in TrickleAppend should probably be changed to `ErrNoRawYet` (or `dm.curNode` should be checked here).

**Comment by kevina** [comment]
> Could you add test cases from this PR: https://github.com/ipfs/go-ipfs/pull/3901/files

I thought the test I added where more comprehensive but I didn't see what you will trying to do.  My added tests conflict with your but will make sure mine are just as comprehensive.

File: unixfs/mod/dagmodifier.go (hunk 2 above)
Comment (kevina) [inline unixfs/mod/dagmodifier.go]: This was not my change. @whyrusleeping care to comment

Comment (whyrusleeping) [inline unixfs/mod/dagmodifier.go]: Yeah, root is probably not the best name here

File: unixfs/mod/dagmodifier.go (hunk 1 above)
Comment (whyrusleeping) [inline unixfs/mod/dagmodifier.go]: fair point. the comment and error are wrong then, maybe we can actually support this part fine now?

Comment (Stebalien) [inline unixfs/mod/dagmodifier.go]: Wait, no; I'm totally wrong. All callers pass `dm.currentNode`. Now, as for why this function even takes a node...

File: unixfs/mod/dagmodifier.go (hunk 2 above)
Comment (kevina) [inline unixfs/mod/dagmodifier.go]: I refactored that method to avoid the problem.

File: unixfs/mod/dagmodifier.go (hunk 1 above)
Comment (kevina) [inline unixfs/mod/dagmodifier.go]: @whyrusleeping I just removed your comment.  If you wish to readd something please give me the correct text to use.

File: test/sharness/t0250-files-api.sh [hunk 4]
```diff
@@ -15,6 +15,7 @@ test_expect_success "can create some files for testing" '
 	FILE1=$(echo foo | ipfs add -q) &&
 	FILE2=$(echo bar | ipfs add -q) &&
 	FILE3=$(echo baz | ipfs add -q) &&
+        FILE9=$(echo zip | ipfs add -q --raw-leaves) &&
```
Comment (Kubuxu) [inline test/sharness/t0250-files-api.sh]: Indentation is different here than in the rest of the file

Comment (kevina) [inline test/sharness/t0250-files-api.sh]: Yeah sorry about that.  Tab vs space issue.

**Comment by kevina** [comment]
@Kubuxu I redid the tests so they are closer to what you had

Raw nodes are now working in the files API as far as I can tell.

**Comment by kevina** [comment]
@whyrusleeping 

> Code LGTM, i would like to see another test in dagmodifier_test.go that tests dagmodifier.Size() on a raw node though

It was non-obvious how to do this so I instead tested that the size was reported correctly in a sharness test.

**Comment by whyrusleeping** [comment]
@kevina The current dagmodifier tests use: https://github.com/ipfs/go-ipfs/blob/master/unixfs/test/utils.go#L45 which generates a random file of the given size, using a standard importer. You could make a new method there that creates a random file using the raw leaves importer option instead, and run some of the same tests on that node type in dagmodifier_test.go

**Comment by kevina** [comment]
@whyrusleeping most all of the tests in dagmodifier_test has to do with appending data, of which is not supported by raw leaves.  It was non-obvious how to modify the test to support the creation of a node without appending.  I can attempt (with some effort) to write a new set of tests for just raw leaves if you want.  

As an aside, I am not sure of the benefit of supporting the full range of operations on raw leaves.  Appending to a unixfs dir node is useful to add files.  How is appending to a raw node useful?  Is there even an API for appending to a file (I don't see it).  Also when appending, what should we do when it becomes too large.

**Comment by Kubuxu** [comment]
@kevina yes there is an API (using `ipfs files write`), when it is too large it should probably be transformed into unixfs dag.

**Comment by kevina** [comment]
@Kubuxu, oh okay, append would be more complicated though and I think should be pushed to a separate p.r. With this one raw nodes are usable within the files api as long as you don't try to create one within the files API.

**Comment by whyrusleeping** [comment]
@kevina thats fair enough. I think we can live with that restriction for now.

**Comment by kevina** [comment]
@whyrusleeping are you okay with testing the size via the sharness tests only?  The existing tests are designed for when the node is created within the API and I will have to write a new set of go tests to test that the raw node size is reported correctly, of which I am unsure how to do so I will have to spend some time figuring it out.

**Comment by whyrusleeping** [comment]
@kevina yeah, for now i guess testing through the api is good enough.

**Comment by kevina** [comment]
@whyrusleeping then I think this p.r. is ready

**Comment by whyrusleeping** [comment]
@kevina mind rebasing on top of latest master? Should make the tests green again

**Pull request merged**